### PR TITLE
Adding GradCache (cached contrastive loss)

### PR DESCRIPTION
**Pull request by NohTow**
This PR adds CachedContrastive loss to the available loss.
As described in the [paper](https://arxiv.org/pdf/2101.06983), this effectively allows to scale the batch size of contrastive loss without requiring more memory (at the cost of a slightly slower training) by decomposing the loss computation into representations computation without the gradient to compute similarities over the whole batch and then computing the gradients on sub batches.

The code is vastly copied from sentence-transformers implementation (thanks @tomaarsen), one of the main difference being the chunking of the score computation as well (as MaxSim is pretty memory intensive for very large batches).

### CODE REVIEW

File: pylate/losses/cached_contrastive.py [hunk 1]
```diff
@@ -0,0 +1,348 @@
...
+        A PyLate ColBERT model
+    score_metric : Callable
+        ColBERT scoring function. Defaults to colbert_scores.
+    mini_batch_size : int
+        Chunk size for the forward pass. You can keep this small to avoid OOM on large batch sizes.
+    size_average : bool
+        Whether to average or sum the cross-entropy loss across the mini-batch.
+    show_progress_bar : bool
+        Whether to show a TQDM progress bar for the embedding steps.
+    Examples
```
Comment (raphaelsty) [inline pylate/losses/cached_contrastive.py:93]: Add space between arguments definitions and unit tests examples

File: pylate/losses/cached_contrastive.py [hunk 2]
```diff
@@ -0,0 +1,348 @@
...
+        loss = self.calculate_loss(reps, masks, with_backward=True)
+        loss = loss.detach().requires_grad_()
+
+        self.cache = [
+            [r.grad for r in rs] for rs in reps
+        ]  # e.g. 3 * bsz/mbsz * (mbsz, hdim)
+
+        return loss
+
+    def calculate_loss(self, reps, masks, with_backward: bool = False) -> Tensor:
```
Comment (raphaelsty) [inline pylate/losses/cached_contrastive.py:220]: Add typing and docstring with arguments definition so it's easier to follow

**Comment by raphaelsty** [comment]
LGTM, this feature is amazing and was not straightforward to adapt

Note, we must be careful with the docstrings otherwise they won't be properly parsed by the system. Here is the standard we defined where the function / class method definition starts on the same line as the quotes. There is a space between the set of parameters and examples

```python
"""Computes the ColBERT scores between queries and documents embeddings. The score is computed as the sum of maximum similarities
    between the query and the document.

    Parameters
    ----------
    queries_embeddings
        The first tensor. The queries embeddings. Shape: (batch_size, num tokens queries, embedding_size)
    documents_embeddings
        The second tensor. The documents embeddings. Shape: (batch_size, num tokens documents, embedding_size)

    Examples
    --------
    >>> import torch

    >>> queries_embeddings = torch.tensor([
    ...     [[1.], [0.], [0.], [0.]],
    ...     [[0.], [2.], [0.], [0.]],
    ...     [[0.], [0.], [3.], [0.]],
    ... ])

    >>> documents_embeddings = torch.tensor([
    ...     [[10.], [0.], [1.]],
    ...     [[0.], [100.], [1.]],
    ...     [[1.], [0.], [1000.]],
    ... ])

    >>> scores = colbert_scores(
    ...     queries_embeddings=queries_embeddings,
    ...     documents_embeddings=documents_embeddings
    ... )

    >>> scores
    tensor([[  10.,  100., 1000.],
            [  20.,  200., 2000.],
            [  30.,  300., 3000.]])

    """
```

File: pylate/losses/cached_contrastive.py [hunk 3]
```diff
@@ -0,0 +1,348 @@
...
+        self._fork.__exit__(exc_type, exc_val, exc_tb)
+        self._fork = None
+
+
+def _backward_hook(
+    grad_output: Tensor,
+    sentence_features: Iterable[dict[str, Tensor]],
+    loss_obj,
+) -> None:
+    """
```
Comment (raphaelsty) [inline pylate/losses/cached_contrastive.py:47]: Description should start on the same line as the quote """ for documentation parsing

File: pylate/losses/cached_contrastive.py [hunk 4]
```diff
@@ -0,0 +1,348 @@
...
+                # Dot the embedding chunk with the cached gradient chunk,
+                # scaled by grad_output from the top-level backward pass
+                surrogate = (
+                    torch.dot(reps_mb.flatten(), grad_mb.flatten()) * grad_output
+                )
+                surrogate.backward()
+
+
+class CachedContrastive(nn.Module):
+    """
```
Comment (raphaelsty) [inline pylate/losses/cached_contrastive.py:75]: Description should start on the same line as the quote """ for documentation parsing

File: pylate/losses/cached_contrastive.py [hunk 5]
```diff
@@ -0,0 +1,348 @@
...
+
+        self.cache = [
+            [r.grad for r in rs] for rs in reps
+        ]  # e.g. 3 * bsz/mbsz * (mbsz, hdim)
+
+        return loss
+
+    def calculate_loss(self, reps, masks, with_backward: bool = False) -> Tensor:
+        """Calculate the cross-entropy loss. No need to cache the gradients."""
+        # Each sub-list in reps is a list of mini-batch chunk embeddings
```
Comment (raphaelsty) [inline pylate/losses/cached_contrastive.py:222]: You could add those comments to the docstring or to the parameters definitions so it's clear from our documentation, the user would not have to dig the code to find those informations

File: pylate/losses/cached_contrastive.py [hunk 6]
```diff
@@ -0,0 +1,348 @@
...
+from torch import Tensor, nn
+from torch.utils.checkpoint import get_device_states, set_device_states
+
+from ..models import ColBERT
+from ..scores import colbert_scores
+from .contrastive import extract_skiplist_mask
+
+
+class RandContext:
+    """
```
Comment (raphaelsty) [inline pylate/losses/cached_contrastive.py:20]: Description should start on the same line as the quote """ for documentation parsing

**Comment by raphaelsty** [comment]
LGTM merging 🥳🐐😎
